File: ignivault/ignivault.API/Security/Crypt.cs
using System.Security.Cryptography;

namespace ignivault.API.Security
{
    public class Crypt
    {
        private const int KeySize = 32;
        private const int SaltSize = 16;
        private const int Iterations = 100_000;

        // Generate a new random key per user (for encrypting vault items)
        public byte[] GenerateRandomKey()
        {
            byte[] key = new byte[KeySize];
            using var rng = RandomNumberGenerator.Create();
            rng.GetBytes(key);
            return key;
        }

        // Generate salt for PBKDF2
        public byte[] GenerateSalt()
        {
            byte[] salt = new byte[SaltSize];
            using var rng = RandomNumberGenerator.Create();
            rng.GetBytes(salt);
            return salt;
        }

        // Derive key from password + salt
        public byte[] DeriveKey(string password, byte[] salt)
        {
            using var deriveBytes = new Rfc2898DeriveBytes(password, salt, Iterations, HashAlgorithmName.SHA256);
            return deriveBytes.GetBytes(KeySize);
        }

        // Encrypt data with AES256 (returns ciphertext + IV)
        public (byte[] Ciphertext, byte[] IV) Encrypt(byte[] plaintext, byte[] key)
        {
            using var aes = Aes.Create();
            aes.KeySize = 256;
            aes.Mode = CipherMode.CBC;
            aes.Padding = PaddingMode.PKCS7;

            aes.GenerateIV();

            using var encryptor = aes.CreateEncryptor(key, aes.IV);
            byte[] ciphertext = encryptor.TransformFinalBlock(plaintext, 0, plaintext.Length);

            return (ciphertext, aes.IV);
        }

        // Decrypt data with AES256
        public byte[] Decrypt(byte[] ciphertext, byte[] key, byte[] iv)
        {
            using var aes = Aes.Create();
            aes.KeySize = 256;
            aes.Mode = CipherMode.CBC;
            aes.Padding = PaddingMode.PKCS7;

            aes.IV = iv;

            using var decryptor = aes.CreateDecryptor(key, aes.IV);
            return decryptor.TransformFinalBlock(ciphertext, 0, ciphertext.Length);
        }

        // Encrypt the master key with a password-derived key
        public (byte[] EncryptedMasterKey, byte[] IV) EncryptMasterKey(byte[] masterKey, string password, byte[] salt)
        {
            byte[] passwordKey = DeriveKey(password, salt);
            return Encrypt(masterKey, passwordKey);
        }

        // Decrypt the master key
        public byte[] DecryptMasterKey(byte[] encryptedMasterKey, string password, byte[] salt, byte[] iv)
        {
            byte[] passwordKey = DeriveKey(password, salt);
            return Decrypt(encryptedMasterKey, passwordKey, iv);
        }
    }
}

File: ignivault/ignivault/wwwroot/js/Crypt.js
window.Crypt = {
    // --- Random bytes -------------------------------------------------------
    generateRandomKey() {
        // 256-bit AES key
        const key = new Uint8Array(32);
        crypto.getRandomValues(key);
        return key;
    },

    generateSalt() {
        // 128-bit salt
        const salt = new Uint8Array(16);
        crypto.getRandomValues(salt);
        return salt;
    },

    // --- Derive key from password + salt ------------------------------------
    async deriveKey(password, salt) {
        // salt: Uint8Array
        const enc = new TextEncoder();
        const keyMaterial = await crypto.subtle.importKey(
            "raw",
            enc.encode(password),
            "PBKDF2",
            false,
            ["deriveKey"]
        );

        const key = await crypto.subtle.deriveKey(
            {
                name: "PBKDF2",
                salt,
                iterations: 100_000,
                hash: "SHA-256"
            },
            keyMaterial,
            { name: "AES-CBC", length: 256 },
            true,
            ["encrypt", "decrypt"]
        );

        return new Uint8Array(await crypto.subtle.exportKey("raw", key));
    },

    // --- AES-CBC Encrypt/Decrypt -------------------------------------------
    async encrypt(plaintextBytes, keyBytes) {
        // plaintextBytes, keyBytes: Uint8Array
        const iv = crypto.getRandomValues(new Uint8Array(16));
        const key = await crypto.subtle.importKey("raw", keyBytes, "AES-CBC", false, ["encrypt"]);

        const ciphertext = await crypto.subtle.encrypt(
            { name: "AES-CBC", iv },
            key,
            plaintextBytes
        );

        return { ciphertext: new Uint8Array(ciphertext), iv };
    },

    async decrypt(ciphertextBytes, keyBytes, iv) {
        const key = await crypto.subtle.importKey("raw", keyBytes, "AES-CBC", false, ["decrypt"]);

        const plaintext = await crypto.subtle.decrypt(
            { name: "AES-CBC", iv },
            key,
            ciphertextBytes
        );

        return new Uint8Array(plaintext);
    },

    // --- Master-key helpers -------------------------------------------------
    async encryptMasterKey(masterKeyBytes, password, salt) {
        const pwdKey = await this.deriveKey(password, salt);
        return this.encrypt(masterKeyBytes, pwdKey);
    },

    async decryptMasterKey(encryptedMasterKeyBytes, password, salt, iv) {
        const pwdKey = await this.deriveKey(password, salt);
        return this.decrypt(encryptedMasterKeyBytes, pwdKey, iv);
    }
};